tallycash/extension
# 🤖Show Me The Data: Introducing the analytics service

# Show Me The Data!

![CleanShot 2022-11-04 at 17 03 42](https://user-images.githubusercontent.com/7690112/200022128-b111e0ee-6d94-4a6a-839b-911dc5ce4ea1.png)
![CleanShot 2022-11-04 at 17 02 19](https://user-images.githubusercontent.com/7690112/200021775-e3fdc3ec-494a-4d3a-8b84-ef6a6b37b6a5.png)

## What 

This PR introduces the analytics service. Implemented events
- background start 
- extension install
- address on network added for tracking 
- UI opened

Not implemented functionality 
- website-related functionality
  - uninstall tracking: It is blocked on a goodbye website page 
  - UTM code tracking with the website: blocked on privacy policy and cookie usage discussion
- data deletion: blocked on tech discussion 
- swaps tracking 
- dApp connections tracking
- signing tracking
- analytics user notification
- wallet usage time > length of time the wallet was open

Note: Tests will be in a follow-up PR.

## How 

Implemented an analytics service, which will be the background connection point. It's responsible for stateful information handling — like checking if the service is turned on or storing/generating the user's uuid — and listening for events. 

The preferences for the analytics service are stored in the preferences service. The data flow is a bit complex because we need to sync the redux and the background state. The settings in the preferences service are handled as the source of truth. This is propagated to the UI on startup and when change happens. Redux state is only updated when the background state is known to be good. 

The posthog-specific concerns are put in a lib, collecting every service-specific implementation detail.

The `USE_ANALYTICS_SOURCE` is meant to signal the context the analytics event was sent from. We can use this field to filter events only interesting to us. You can overwrite it with whatever you want in your local environment. This will help find only events that originated from your setup. 

## Test

Note: ⚠️ The analytics is turned off by default. Testing is more straightforward if you set [`defaultPreferences.analytics.isEnabled`](https://github.com/tallycash/extension/blob/7936d5fce3b2722bf5a0306c92397286dd884a91/background/services/preferences/defaults.ts#L33) to `true`.

- download the asset or change the environment variables and build locally 
- import a seed phrase 
- go to settings and turn on the feature
- reload the extension
- wait 1+ min 
- check the uuid in devtools > Application > indexedDB > `tally/analytics`
- go to https://app.posthog.com/events > Live events
  - if you need an account reach out to me or @puppycodes 
- you should see`New Install` and a `Bakckground start` event 
- open the extension: `UI open` should arrive

## Testing Env

```
SUPPORT_ANALYTICS=true
SHOW_ANALYTIC_MENU=true
SHOW_DELETE_DATA_BUTTON=false
```

Latest build: [extension-builds-2546](https://github.com/tallycash/extension/suites/9338393780/artifacts/439117915) (as of Wed, 16 Nov 2022 13:37:33 GMT).
